packages/react-dev-overlay/src/middleware-turbopack.ts

```diff
@@ -9,57 +9,88 @@ import { codeFrameColumns } from '@babel/code-frame'
 import { launchEditor } from './internal/helpers/launchEditor'
 
 interface Project {
-  traceSource(stackFrame: RustStackFrame): Promise<
-    | {
-        frame: RustStackFrame
-        source: string
-      }
-    | undefined
-  >
+  getSourceForAsset(filePath: string): Promise<string | null>
+  traceSource(stackFrame: RustStackFrame): Promise<RustStackFrame | null>
 }
 
 interface RustStackFrame {
   file: string
-  methodName: string | undefined
+  methodName: string | null
   line: number
-  column: number | undefined
+  column: number | null
 }
 
-export async function createOriginalStackFrame(
-  project: Project,
-  frame: StackFrame
-): Promise<OriginalStackFrameResponse | null> {
-  const source = await project.traceSource({
-    file: frame.file ?? '<unknown>',
+const currentSourcesByFile: Map<string, Promise<string | null>> = new Map()
+async function batchedTraceSource(project: Project, frame: StackFrame) {
+  const file = frame.file
+  if (!file) {
+    return
+  }
+
+  const rustStackFrame = {
+    file,
     methodName: frame.methodName,
     line: frame.lineNumber ?? 0,
-    column: frame.column ?? undefined,
-  })
+    column: frame.column,
+  }
 
-  if (!source) {
-    return null
+  const sourceFrame = await project.traceSource(rustStackFrame)
+  if (!sourceFrame) {
+    return
+  }
+
+  let source
+  // Don't show code frames for node_modules. These can also often be large bundled files.
+  if (!sourceFrame.file.includes('node_modules')) {
+    let sourcePromise = currentSourcesByFile.get(sourceFrame.file)
+    if (!sourcePromise) {
+      sourcePromise = new Promise((resolve) =>
+        // Batch reading sources content as this can be quite large, and stacks often reference the same files
+        setTimeout(resolve, 100)
```

Comment:
I think we could change this to immediately execute, and keep in in memory for 100ms after it resolves? That way we don't add a 100ms delay to the actual tracing.